File: backend/workflows/states.py
from typing import Any, TypedDict


class IngestState(TypedDict, total=False):
    video_id: str
    status: str
    transcript: list[dict]
    chunks: list[dict]
    total_chunks: int
    processed_chunks: int
    error: str | None
    completed: bool


class QueryState(TypedDict, total=False):
    video_id: str
    query: str
    chat_history: list
    cached_summary: str | None
    filter_message: str | None
    intent: str
    search_query: str
    context: str
    response: str
    response_stream: Any
    summary_generated: bool
    error: str | None

File: frontend/src/lib/rag/workflows/states.ts
import { Annotation } from "@langchain/langgraph";
import type { ChatHistoryMessage, TextChunk, TranscriptSegment } from "../types";

export const QueryStateAnnotation = Annotation.Root({
  videoId: Annotation<string>,
  query: Annotation<string>,
  chatHistory: Annotation<ChatHistoryMessage[]>,
  cachedSummary: Annotation<string | null>,
  filterMessage: Annotation<string | null>,
  intent: Annotation<"SUMMARY" | "QA">,
  searchQuery: Annotation<string>,
  context: Annotation<string | null>,
  response: Annotation<string>,
  summaryGenerated: Annotation<boolean>,
  error: Annotation<string | null>,
});

export type QueryState = typeof QueryStateAnnotation.State;

export const IngestStateAnnotation = Annotation.Root({
  videoId: Annotation<string>,
  status: Annotation<string>,
  transcript: Annotation<TranscriptSegment[]>,
  chunks: Annotation<TextChunk[]>,
  totalChunks: Annotation<number>,
  processedChunks: Annotation<number>,
  error: Annotation<string | null>,
  completed: Annotation<boolean>,
});

export type IngestState = typeof IngestStateAnnotation.State;
